# Resampling Methods (Introduction)
* *Resampling methods* is an indispensable tool in modern statistics, involve repeatedly drawing samples from a training set and refitting a model of interest on each sample in order to obtain additional information about the fitted model.
* This approach may allow us to obtain information that would not be available from fitting the model only once using the original training sample.
* Can be computationally expensive (involve fitting the same statistical method multiple times using different subsets of the training data).
* Commonly used resampling methods - `Cross Validation` and `Bootstraping`.
* ***Cross-validation*** can be used to estimate the test error associated with a given statistical learning method in order to evaluate its performance, or to select the appropriate level of flexibility :
  * The process of evaluating a model’s performance is known as *model assessment*.
  * The process of selecting the proper level of flexibility for a model is known as *model selection*.
* On the other hand, ***bootstrap***is to provide a measure of accuracy of a parameter estimate or of a given statistical learning method.


# Cross-validation
* The *test error* is the average error that results from using a statistical learning method to predict the response on a new observation - can be easily calculated if a designated test set is available - Unfortunately, it is not the case in reality -  computed as $$ test error = Avg(y_{0}- \hat{f}(x_{0}))^2$$
* The *training error* using the training data that was used to fit the model, and so should more accurately and it is computed as
$$ MSE_{train}=\frac{1}{n}\cdot \sum_{i=1}^{n}\left(y_{i}-\hat{f}(x_{i})\right)^2 $$
* We want to choose the method that gives the ***lowest test MSE***, as opposed to the *lowest training MSE*.
* There is no guarantee that the method with the lowest training MSE will also have the lowest test MSE.
* Goal is to find suitable means for calculating *test error* in the absence of a very large designated test set that can be used to directly estimate the test error rate!
### The Validation Set Approach
* It involves randomly dividing the available set of observations into two parts, a *training set* and a *validation set* a.k.a *hold-out set*.
* Now, model is fit on the training set, and the fitted model is used to predict the responses for the observations in the validation set.
* The resulting validation set error rate provides an estimate of the test error rate.
* If we repeat the process of randomly splitting the sample set into two parts, we will get a somewhat different estimate for the test MSE.
* The validation set approach is conceptually simple and is easy to implement.
* Drawbacks of Validation Set Approach -
  * The validation estimate of the test error rate can be highly variable, depending upon precisely which observations are included in the training set and validation sets respectively.
  * Since, only a subset of the observations are made as training set and validation sets, as per statistical methods tend to perform worse when trained on fewer observations, validation set error rate may tend to **over-estimate** the test error rate for the model fit on the entire data set.
### Leave-One-Out Cross-Validation (LOOCV)
* LOOCV involves splitting the set of observations into two parts - a single observation (x1, y1) is used for the validation set, and the remaining observations {(x2, y2), . . . , (xn, yn)} make up the training set.
* The statistical learning method is fit on the n − 1 training observations, and a prediction is made for the excluded observation, using its value x1.
* MSE will be calculated as MSE = (y1-ŷ)^2. But even though MSE1 is unbiased for the test error, it is a poor estimate because it is highly variable, since it is based upon a single observation (x1, y1).
* Repeating this approach *n* times produces *n* squared errors,
  $$CV_{(n)} = \frac{1}{n}\cdot \sum_{i=1}^{n} MSE_{i}$$.
* LOOCV has the potential to be expensive to implement, since the model has to be fit *n* times.
* But an amazing shortcut makes the cost of LOOCV the same as that of a single model fit! The following formula holds:
$$CV_{(n)} = \frac{1}{n}\cdot \sum_{i=1}^{n}\left(\frac{y_{i}-\hat{y}}{1-h_{i}}\right)^2$$.

### k-Fold Cross-Validation
* Alternative to LOOCV.
* This approach involves randomly dividing the set of observations into *k* groups, or *folds*, of approximately equal size.
* The first fold is treated as a validation set, and the method is fit on the remaining k − 1 folds.
* The mean squared error, MSE1, is then computed on the observations in the *held-out* fold.
* This procedure is repeated *k* times; each time, a different group of observations is treated as a validation set.
* This process results in k estimates of the test error, k-fold CV estimate is computed.
* LOOCV is a special case of k-fold CV in which k is set to equal n (In practice k=5 or k=10 is performed).
* When we perform cross-validation, our goal might be to determine how well a given statistical learning procedure can be expected to perform on independent data - the actual estimate of the *test MSE*.
* But at other times, ***location of the minimum point in the estimated test MSE curve*** - we might be performing cross-validation on a number of statistical learning methods, or on a single method using different levels of flexibility, in order to identify the method that results in the lowest test error.
* An important advantage of k-fold CV - it often gives more accurate estimates of the test error rate than does LOOCV.
* From the perspective of bias reduction, it is clear that LOOCV is to be preferred to k-fold CV. ***Note***: LOOCV will give approximately unbiased estimates of the test error, since each training set contains n−1 observations and performing k-fold CV for, say, k = 5 or k = 10 will lead to an intermediate level of bias.
* But variance perspective also needs to taken into consideration! LOOCV has higher variance than does k-fold CV with k < n. When we perform LOOCV, we are in effect averaging the outputs of n fitted models, each of which is trained on an almost identical set of observations; therefore, these outputs are highly (positively) correlated
with each other. When we perform k-fold CV with k < n, we are averaging the outputs of k fitted models that are somewhat less correlated with each other, since the overlap between the training sets in each model is smaller.
* The test error estimate resulting from LOOCV tends to have higher variance than does the test error estimate resulting from k-fold CV.
* To summarize, there is a bias-variance trade-off associated with the choice of k in k-fold cross-validation.
 $$CV_{(k)} = \frac{1}{k}\cdot \sum_{i=1}^{k} MSE_{i}$$.
### Cross-Validation on Classification Problems
* Cross-validation can also be a very useful approach in the classification setting when Y is *qualitative*.
* Rather than using MSE to quantify test error, we instead use the ***number of misclassified observations***.
* the training error tends to ↓, as the flexibility of the fit ↑.

## Bootstrap
* Widely applicable and extremely powerful statistical tool that can be used to quantify the uncertainty associated with a given estimator or statistical learning method.
* The power of the bootstrap lies in the fact that it can be easily applied to a wide range of statistical learning methods, including some for which a measure of variability is otherwise difficult to obtain.
* The bootstrap approach allows us to use a computer to emulate the process of obtaining new sample sets.
* Rather than repeatedly obtaining independent data sets from the population, we instead obtain distinct data sets by repeatedly sampling observations from the original data set.
* When the sampling is performed with replacement, which means that the same observation can occur more than once in the bootstrap data set, we call them ***sampling with replacement*** and vice-versa.
<br>
Let `Z`→ original data set with `n` observations.<br> We randomly take n observations from the data set in order to produce a bootstrap data set `Z*1`.<br> The sampling is performed ***with replacement*** and thus its estimate will be `^α1`.<br> Let `B` → no.of.times bootstrapped; for some large value of B. <br> Thus, corresponding `α`
estimates, `ˆα1, ˆα2, . . . , ˆαB`. <br> Then the *estimate of the standard error of `^α` is given by: * $$ SE_{B}(\hat{\alpha}) = \sqrt{\frac{1}{B-1}\cdot \sum_{r=1}^{B}\left(\hat{\alpha}^\ast r-\frac{1}{B}\sum_{r'=1}^{B} \hat{\alpha}^\ast r'\right)^2}$$

# Lab Codes: Cross-Validation and the Bootstrap

## Step 1:Imports

In [1]:
import numpy as np
import statsmodels.api as sm
from ISLP import load_data
from ISLP.models import (ModelSpec as MS,
summarize ,
poly)
from sklearn.model_selection import train_test_split

## Step 2: Neccesary modules import for CV and Bootstrappings

In [2]:
from functools import partial
from sklearn.model_selection import \
(cross_validate ,
KFold ,
ShuffleSplit)
from sklearn.base import clone
from ISLP.models import sklearn_sm

## Step 3A:  The Validation Set Approach
* We explore the use of the *validation set approach* in order to estimate the test error rates that result from fitting various linear models on the `Auto` data set.
* We use the function `train_test_split()` to split the data into training and validation sets.
* As there are 392 observations, we split into two equal sets of size 196 using the argument `test_size=196`.
* It is generally a good idea to set a *random seed* when performing operations like this that contain an element of randomness, so that the results obtained can be reproduced precisely at a later time.
* We set the random seed of the splitter with the argument `random_state=0`.

In [38]:
Auto = load_data('Auto')
Auto_train , Auto_valid = train_test_split(Auto ,
                                           test_size=196,
                                           random_state=0)

Now we can fit a *linear regression* using only the observations corresponding to the training set `Auto_train`.

In [4]:
hp_mm = MS(['horsepower'])
X_train = hp_mm.fit_transform(Auto_train)
y_train = Auto_train['mpg']
model = sm.OLS(y_train , X_train)
results = model.fit()

We now use the `predict()` method of `results` evaluated on the model matrix for this model created using the validation data set. We also calculate the validation MSE of our model.

In [5]:
X_valid = hp_mm.transform(Auto_valid)
y_valid = Auto_valid['mpg']
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2)

np.float64(23.61661706966988)

Hence our estimate for the validation MSE of the linear regression fit is 23.62.<br>
We can also estimate the validation error for higher-degree polynomial regressions. 
We first provide a function `evalMSE()` that takes a model string as well as a training and test set and returns the MSE on the test set.

In [6]:
def evalMSE(terms ,response ,train ,test):
    mm = MS(terms)
    X_train = mm.fit_transform(train)
    y_train = train[response]
    X_test = mm.transform(test)
    y_test = test[response]
    results = sm.OLS(y_train , X_train).fit()
    test_pred = results.predict(X_test)
    return np.mean((y_test - test_pred)**2)

Let’s use this function to estimate the validation MSE using linear, quadratic and cubic fits. <br>We use the `enumerate()` function here, which gives both the values and indices of objects as one iterates over a for loop.

In [7]:
MSE = np.zeros(3)
for idx , degree in enumerate(range(1, 4)):
    MSE[idx] = evalMSE([poly('horsepower', degree)],'mpg',Auto_train ,Auto_valid)
MSE

array([23.61661707, 18.76303135, 18.79694163])

These error rates are 23.62, 18.76, and 18.80, respectively. If we choose a different training/validation split instead, then we can expect somewhat different errors on the validation set.

In [8]:
Auto_train , Auto_valid = train_test_split(Auto , test_size=196, random_state=3)
MSE = np.zeros(3)
for idx , degree in enumerate(range(1, 4)):
    MSE[idx] = evalMSE([poly('horsepower', degree)],'mpg',Auto_train ,Auto_valid)
MSE

array([20.75540796, 16.94510676, 16.97437833])

Using this split of the observations into a training set and a validation set, we find that the validation set error rates for the models with linear, quadratic, and cubic terms are 20.76, 16.95, and 16.97, respectively.<br>
These results are consistent with our previous findings: a model that predicts mpg using a quadratic function of `horsepower` performs better than a model that involves only a linear function of `horsepower`, and there is no evidence of an improvement in using a cubic function of `horsepower`.

## Step 3B: Cross-Validation
* In theory, the cross-validation estimate can be computed for any generalized linear model.
* In practice, however, the simplest way to cross-validate in Python is to use `sklearn`, which has a different interface or API than `statsmodels`, the code we have been using to fit GLMs.
* This is a problem which often confronts data scientists: “*I have a function to do task A, and need to feed it into something that performs task B, so that I can compute B(A(D)), where D is my data.*” When A and B don’t naturally speak to each other, this requires the use of a **wrapper**.
* In the `ISLP` pacakage, we provide a wrapper, `sklearn_sm()`, that enables us to easily use the cross-validation tools of `sklearn` with models fit by `statsmodels`.
* The class `sklearn_sm()` has as its first argument a model from `statsmodels`.
* It can take two additional optional arguments: `model_str` which can be used to specify a formula, and `model_args` which should be a dictionary of additional arguments used when fitting the model.
* For example, to fit a logistic regression model we have to specify a `family` argument.
* This is passed as `model_args={'family':sm.families.Binomial()}`.
* Here is our wrapper in action -

In [9]:
hp_model = sklearn_sm(sm.OLS ,
                      MS(['horsepower']))
X, Y = Auto.drop(columns=['mpg']), Auto['mpg']
cv_results = cross_validate(hp_model ,
                            X,
                            Y,
                            cv=Auto.shape[0])
cv_err = np.mean(cv_results['test_score'])
cv_err

np.float64(24.231513517929216)

1. The arguments to `cross_validate()` are as follows:
   * An object with the appropriate `fit()`, `predict()`, and `score()` methods, an array of features X and a response Y.
2. We also included an additional argument `cv` to `cross_validate()`; specifying an integer `K` results in *K-fold cross-validation*.
3. We have provided a value corresponding to the total number of observations, which results in *leave-one-out cross-validation (LOOCV)*.
4. The `cross_validate()` function produces a dictionary with several components; we simply want the cross validated test score here (MSE), which is estimated to be 24.23.<br>
* We can repeat this procedure for increasingly complex polynomial fits.
* To automate the process, we again use a `for` loop which iteratively fits polynomial regressions of degree 1 to 5, computes the associated crossvalidation error, and stores it in the ith element of the vector `cv_error`.
* The variable `d` in the `for` loop corresponds to the degree of the polynomial.
* We begin by initializing the vector.
* This command may take a couple of seconds to run.

In [10]:
cv_error = np.zeros(5)
H = np.array(Auto['horsepower'])
M = sklearn_sm(sm.OLS)
for i, d in enumerate(range(1,6)):
    X = np.power.outer(H, np.arange(d+1))
    M_CV = cross_validate(M,
                          X,
                          Y,
                          cv=Auto.shape[0])
    cv_error[i] = np.mean(M_CV['test_score'])
cv_error

array([24.23151352, 19.24821312, 19.33498406, 19.42443031, 19.03320903])

We see a sharp drop in the estimated test MSE between the linear and quadratic fits, but then no clear improvement from using higher-degree polynomials.<br>
* Above we introduced the `outer()` method of the `np.power()` function.
* The `outer()` method is applied to an operation that has two arguments, such as `add()`, `min()`, or `power()`.
* It has two arrays as arguments, and then forms a larger array where the operation is applied to each pair of elements of the two arrays.

In [11]:
A = np.array([3, 5, 9])
B = np.array([2, 4])
np.add.outer(A, B)

array([[ 5,  7],
       [ 7,  9],
       [11, 13]])

* In the CV example above, we used `K = n`, but of course we can also use `K < n`.
* The code is very similar to the above (and is significantly faster).
* Here we use `KFold()` to partition the data into `K = 10` random groups.
* We use random_state to set a random seed and initialize a vector `cv_error` in which we will store the CV errors corresponding to the polynomial fits of degrees one to five.

In [12]:
cv_error = np.zeros(5)
cv = KFold(n_splits=10,shuffle=True ,random_state=0) # use same splits for each degree

for i, d in enumerate(range(1,6)):
    X = np.power.outer(H, np.arange(d+1))
    M_CV = cross_validate(M,
                          X,
                          Y,
                          cv=cv)
    cv_error[i] = np.mean(M_CV['test_score'])
cv_error

array([24.20766449, 19.18533142, 19.27626666, 19.47848402, 19.13722633])

1. Notice that the computation time is much shorter than that of LOOCV.(In principle, the computation time for LOOCV for a least squares linear model should be faster than for K-fold CV, due to the availability of the
formulafor LOOCV; however, the generic `cross_validate()` function does not make use of this formula.)
2. We still see little evidence that using cubic or higher-degree polynomial terms leads to a lower test error than simply using a quadratic fit.
3. The `cross_validate()` function is flexible and can take different splitting mechanisms as an argument.
4. For instance, one can use the `ShuffleSplit()` funtion to implement the validation set approach just as easily as K-fold cross-validation.

In [13]:
validation = ShuffleSplit(n_splits=1,
                          test_size=196,
                          random_state=0)
results = cross_validate(hp_model ,
                         Auto.drop(['mpg'], axis=1),
                         Auto['mpg'],
                         cv=validation);
results['test_score']

array([23.61661707])

One can estimate the variability in the test error by running the following-

In [14]:
validation = ShuffleSplit(n_splits=10,
                          test_size=196,
                          random_state=0)
results = cross_validate(hp_model ,
                         Auto.drop(['mpg'], axis=1),
                         Auto['mpg'],
                         cv=validation)
results['test_score'].mean(), results['test_score'].std()

(np.float64(23.802232661034164), np.float64(1.4218450941091847))

Note that this *standard deviation* is not a valid estimate of the sampling variability of the *mean* test score or the individual scores, since the randomly-selected training samples overlap and hence introduce correlations. But it does give an idea of the **Monte Carlo variation** incurred by picking different random folds.

## Step 3C: The Bootstrap
### Estimating the Accuracy of a Statistic of Interest
* One of the great advantages of the bootstrap approach is that it can be applied in almost all situations.
* No complicated mathematical calculations are required.
* While there are several implementations of the bootstrap in Python, its use for estimating standard error is simple enough that we write our own function below for the case when our data is stored in a dataframe.
* To illustrate the bootstrap, we start with a simple example.
* The `Portfolio` data set in the `ISLP` package is described in Section 5.2.
* The goal is to estimate the sampling variance of the parameter ) given in formula (5.7).
* We will create a function `alpha_func()`, which takes as input a dataframe `D` assumed to have columns `X` and `Y`, as well as a vector `idx` indicating which observations should be used to estimate `α`.
* The function then outputs the estimate for `α` based on the selected observations.

In [15]:
Portfolio = load_data('Portfolio')
def alpha_func(D, idx):
    cov_ = np.cov(D[['X','Y']].loc[idx], rowvar=False)
    return ((cov_[1,1] - cov_[0,1]) /(cov_[0,0]+cov_[1,1]-2*cov_[0,1]))

This function returns an estimate for `α` based on applying the minimum variance formula to the observations indexed by the argument `idx`. For instance, the following command estimates `α` using all 100 observations.

In [16]:
alpha_func(Portfolio , range(100))

np.float64(0.57583207459283)

Next we randomly select 100 observations from `range(100)`, with replacement. This is equivalent to constructing a new bootstrap data set and recomputing `ˆα` based on the new data set.

In [17]:
rng = np.random.default_rng(0)
alpha_func(Portfolio ,
           rng.choice(100,100,replace=True)
          )

np.float64(0.6074452469619004)

This process can be generalized to create a simple function `boot_SE()` for computing the bootstrap standard error for arbitrary functions that take only a data frame as an argument.

In [61]:
def boot_SE(func, D, n=None, B=1000, seed=0, *args, **kwargs):
    rng = np.random.default_rng(seed)
    first_, second_ = 0, 0 
    n = n or D.shape[0]

    for _ in range(B):
        # Use integer positions instead of index labels
        idx = rng.choice(len(D), n, replace=True)
        value = func(D, idx)
        first_ += value
        second_ += value**2
        
    return np.sqrt(second_ / B - (first_ / B)**2)

Notice the use of `_ `as a loop variable in `for _ in range(B)`. This is often used if the value of the counter is unimportant and simply makes sure the loop is executed `B` times.<br>Let’s use our function to evaluate the accuracy of our estimate of α using B = 1,000 bootstrap replications.

In [19]:
alpha_SE = boot_SE(alpha_func ,
                   Portfolio ,
                   B=1000,
                   seed=0)
alpha_SE

np.float64(0.019199498387420112)

The final output shows that the bootstrap estimate for `SE(ˆα)` is 0.0912.

### Estimating the Accuracy of a Linear Regression Model
* The bootstrap approach can be used to assess the variability of the coefficient estimates and predictions from a statistical learning method.
* Here we use the bootstrap approach in order to assess the variability of the estimates for `β₀` and `β₁` the intercept and slope terms for the linear regression model that uses `horsepower` to predict `mpg` in the `Auto` data set.
* We will compare the estimates obtained using the bootstrap to those obtained using the formulas for `SE(ˆβ₀)` and `SE(ˆβ₁)`.
* To use our `boot_SE()` function, we must write a function (its first argument) that takes a data frame `D` and indices `idx` as its only arguments.
* But here we want to bootstrap a specific regression model, specified by a model formula and data.
* We show how to do this in a few simple steps.
* We start by writing a generic function `boot_OLS()` for bootstrapping a regression model that takes a formula to define the corresponding regression.
* We use the `clone()` function to make a copy of the formula that can be refit to the new dataframe.
* This means that any derived features such as those defined by `poly()` (which we will see shortly), will be re-fit on the resampled data frame.

In [59]:
def boot_OLS(model_matrix , response , D, idx):
    D_ = D.iloc[idx]
    Y_ = D_[response]
    X_ = clone(model_matrix).fit_transform(D_)
    return sm.OLS(Y_, X_).fit().params

* This is not quite what is needed as the first argument to `boot_SE()`.
* The first two arguments which specify the model will not change in the bootstrap process, and we would like to freeze them.
* The function `partial()` from the `functools` module does precisely this: it takes a function as an argument, and freezes some of its arguments, starting from the left.
* We use it to freeze the first two model-formula arguments of `boot_OLS()`.

In [21]:
hp_func = partial(boot_OLS , MS(['horsepower']), 'mpg')

Typing `hp_func?` will show that it has two arguments `D` and `idx` — it is a version of `boot_OLS()` with the first two arguments frozen — and hence is ideal as the first argument for `boot_SE()`. The `hp_func()` function can now be used in order to create bootstrap estimates for the intercept and slope terms by randomly sampling from among the observations with replacement. We first demonstrate its utility on 10 bootstrap samples.

In [22]:
hp_func?

Signature:      hp_func(D, idx)
Call signature: hp_func(*args, **kwargs)
Type:           partial
String form:    functools.partial(<function boot_OLS at 0x00000193A5D25EE0>, ModelSpec(terms=['horsepower']), 'mpg')
File:           c:\users\vijay\appdata\local\programs\python\python311\lib\functools.py
Docstring:     
partial(func, *args, **keywords) - new function with partial application
of the given arguments and keywords.

In [29]:
Auto_reset = Auto.reset_index(drop=True)
rng = np.random.default_rng(0)
np.array([hp_func(Auto_reset,
                  rng.choice(392, 392, replace=True)) 
          for _ in range(10)])

array([[39.88064456, -0.1567849 ],
       [38.73298691, -0.14699495],
       [38.31734657, -0.14442683],
       [39.91446826, -0.15782234],
       [39.43349349, -0.15072702],
       [40.36629857, -0.15912217],
       [39.62334517, -0.15449117],
       [39.0580588 , -0.14952908],
       [38.66688437, -0.14521037],
       [39.64280792, -0.15555698]])

Next, we use the `boot_SE()` function to compute the standard errors of 1,000 bootstrap estimates for the intercept and slope terms.

In [32]:
hp_se = boot_SE(hp_func ,
                Auto ,
                B=1000,
                seed=10)
hp_se

intercept     1.270578
horsepower    0.005293
dtype: float64

In [31]:
hp_model.fit(Auto , Auto['mpg'])
model_se = summarize(hp_model.results_)['std err']
model_se

intercept     0.717
horsepower    0.006
Name: std err, dtype: float64

* The standard error estimates for `ˆβ₀` and `ˆβ₁` obtained using the formulas  are 0.717 for the intercept and 0.006 for the slope.
* Interestingly, these are somewhat different from the estimates obtained using the bootstrap.
* Does this indicate a problem with the bootstrap?
* In fact, it suggests the opposite.
* Recall that the standard formulas, rely on certain assumptions.
* For example, they depend on the unknown parameter `σ²`, the noise variance.
* We then estimate σ² using the RSS.
* Now although the formula for the standard errors do not rely on the linear model being correct, the estimate for σ² does.
* We see that there is a non-linear relationship in the data, and so the residuals from a linear fit will be inflated, and so will `ˆσ²`.
* Secondly, the standard formulas assume (somewhat unrealistically) that the `xi` are fixed, and all the variability comes from the variation in the errors 𝛆ᵢ.
* The bootstrap approach does not rely on any of these assumptions, and so it is likely giving a more accurate estimate of the standard errors of `ˆβ₀` and `ˆβ₁` than the results from `sm.OLS`.
* Below we compute the bootstrap standard error estimates and the standard linear regression estimates that result from fitting the quadratic model to the data.
* Since this model provides a good fit to the data, there is now a better correspondence between the bootstrap estimates and the standard estimates of `SE(ˆβ₀)`, `SE(ˆβ₁)` and `SE(ˆβ₂)`.

In [43]:
Auto = load_data('Auto')

In [44]:
Auto

mpg  cylinders  displacement  horsepower  weight  \
name                                                                           
chevrolet chevelle malibu  18.0          8         307.0         130    3504   
buick skylark 320          15.0          8         350.0         165    3693   
plymouth satellite         18.0          8         318.0         150    3436   
amc rebel sst              16.0          8         304.0         150    3433   
ford torino                17.0          8         302.0         140    3449   
...                         ...        ...           ...         ...     ...   
ford mustang gl            27.0          4         140.0          86    2790   
vw pickup                  44.0          4          97.0          52    2130   
dodge rampage              32.0          4         135.0          84    2295   
ford ranger                28.0          4         120.0          79    2625   
chevy s-10                 31.0          4         119.0          82    2720   

                           acceleration  year  origin  
name                                                   
chevrolet chevelle malibu          12.0    70       1  
buick skylark 320                  11.5    70       1  
plymouth satellite                 11.0    70       1  
amc rebel sst                      12.0    70       1  
ford torino                        10.5    70       1  
...                                 ...   ...     ...  
ford mustang gl                    15.6    82       1  
vw pickup                          24.6    82       2  
dodge rampage                      11.6    82       1  
ford ranger                        18.6    82       1  
chevy s-10                         19.4    82       1  

[392 rows x 8 columns]

In [62]:
quad_model = MS([poly('horsepower', 2, raw=True)])
quad_func = partial(boot_OLS ,
quad_model ,
'mpg')
boot_SE(quad_func , Auto , B=1000)

intercept                                  2.067840
poly(horsepower, degree=2, raw=True)[0]    0.033019
poly(horsepower, degree=2, raw=True)[1]    0.000120
dtype: float64

We compare the results to the standard errors computed using `sm.OLS()`.

In [63]:
M = sm.OLS(Auto['mpg'],
           quad_model.fit_transform(Auto))
summarize(M.fit())['std err']

intercept                                  1.800
poly(horsepower, degree=2, raw=True)[0]    0.031
poly(horsepower, degree=2, raw=True)[1]    0.000
Name: std err, dtype: float64